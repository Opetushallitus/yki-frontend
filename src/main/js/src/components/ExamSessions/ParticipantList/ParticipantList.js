import { parsePhoneNumberFromString } from 'libphonenumber-js';
import moment from 'moment';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import React, { useState } from 'react';
import { withTranslation } from 'react-i18next';

import checkMarkDone from '../../../assets/svg/checkmark-done.svg';
import checkMarkNotDone from '../../../assets/svg/checkmark-not-done.svg';
import trashcan from '../../../assets/svg/trashcan.svg';
import { DATE_FORMAT } from '../../../common/Constants';
import { ActionButton } from '../../UI/ActionButton/ActionButton';
import ListExport from './ListExport/ListExport';
import RelocateParticipant from './RelocateParticipant/RelocateParticipant';
import classes from './ParticipantList.module.css';
import { examSessionParticipantsCount } from '../../../util/examSessionUtil';

const stateComparator = () => (a, b) => {
  if (a.state === 'COMPLETED') return -1;
  if (b.state === 'COMPLETED') return 1;
  if (a.state === 'SUBMITTED') return -1;
  if (b.state === 'SUBMITTED') return 1;

  return 0;
};

const sortByNames = () => R.sortWith([
  R.ascend(R.path(['form', 'last_name'])),
  R.ascend(R.path(['form', 'first_name'])),
]);

export const participantList = props => {
  const [sortParticipantsFn, setSortParticipantsFn] = useState(sortByNames);

  const getStateTranslationKey = state => {
    switch (state) {
      case 'COMPLETED':
        return 'examSession.paid';
      case 'CANCELLED':
        return 'examSession.cancelled';
      case 'EXPIRED':
        return 'examSession.expired';
      case 'PAID_AND_CANCELLED':
        return 'examSession.paidAndCancelled';
      default:
        return 'examSession.notPaid';
    }
  };

  const registrationStatus = participant => {
    const registrationState = participant.state;
    const image =
      registrationState === 'COMPLETED' ? checkMarkDone : checkMarkNotDone;
    const text = props.t(getStateTranslationKey(registrationState));

    return (
      <React.Fragment>
        <img src={image} data-cy={`registration-${registrationState}`} alt="" />{' '}
        {text}
      </React.Fragment>
    );
  };

  const ssnOrBirthDate = form => {
    return form.ssn ? form.ssn : moment(form.birthdate).format(DATE_FORMAT);
  };

  const getPhoneNumber = participant => {
    const asNumber = parsePhoneNumberFromString(participant.form.phone_number);
    return asNumber ? asNumber.formatInternational() : '';
  };

  const relocateParticipant = participant => {
    return (
      <RelocateParticipant
        examSession={props.examSession}
        examSessions={props.examSessions}
        onRelocate={newSessionId =>
          props.onRelocate(
            props.examSession.organizer_oid,
            props.examSession.id,
            participant.registration_id,
            newSessionId,
          )
        }
        confirmText={props.t('examSession.registration.relocate.confirm')}
      />
    );
  };

  const handleFilterChange = event => {
    switch (event.target.value) {
      case 'name':
        setSortParticipantsFn(sortByNames);
        break;
      case 'state':
        setSortParticipantsFn(() => R.sort(stateComparator()));
        break;
      case 'registrationTime':
        setSortParticipantsFn(() => R.sortBy(R.prop('created')));
        break;
      case 'registrationType':
        setSortParticipantsFn(() => R.sortBy(R.prop('kind')));
        break;
      default:
        setSortParticipantsFn(() => R.sortBy(R.prop('created')));
        break;
    }
  };

  const participantFiltering = () => {
    return (
      <>
        <label htmlFor="participantFilter">
          {props.t('examSession.participants.sortBy')}
        </label>
        <select
          id="ParticipantFilter"
          className={classes.ParticipantFilter}
          onChange={handleFilterChange}
        >
          <option value="name">
            {props.t('examSession.participants.sortBy.name')}
          </option>
          <option value="registrationTime">
            {props.t('examSession.participants.sortBy.registrationTime')}
          </option>
          <option value="registrationType">
            {props.t('examSession.participants.sortBy.registrationType')}
          </option>
          <option value="state">
            {props.t('examSession.participants.sortBy.state')}
          </option>
        </select>
      </>
    );
  };

  const cancelRegistrationButton = p => {
    const cancelRegistration = (
      <React.Fragment>
        <img src={trashcan} alt="" />{' '}
        {props.t('examSession.registration.cancel')}
      </React.Fragment>
    );
    return (
      <ActionButton
        children={cancelRegistration}
        confirmOnRight={true}
        onClick={() =>
          props.onCancel(
            props.examSession.organizer_oid,
            props.examSession.id,
            p.registration_id,
          )
        }
        confirmText={props.t('examSession.registration.cancel.confirm')}
        cancelText={props.t('examSession.registration.cancel.cancel')}
      />
    );
  };

  const participantRows = participants => {
    return sortParticipantsFn(participants).map((p, i) => (
      <React.Fragment key={i}>
        <div
          className={[
            classes.ItemHeader,
            classes.Index,
            classes.StateItem,
          ].join(' ')}
          data-cy={`participant-${p.registration_id}`}
        >
          {i + 1}.
        </div>
        <div className={[classes.ItemHeader, classes.StateItem].join(' ')}>
          {p.form.first_name} {p.form.last_name}
        </div>
        <div
          className={[
            classes.ItemHeader,
            p.state === 'COMPLETED' ? classes.StatusCompleted : classes.Status,
            classes.StateItem,
          ].join(' ')}
        >
          {registrationStatus(p)}
        </div>
        <div className={classes.StateItem}>
          {p.created && moment(p.created).format(DATE_FORMAT)}
        </div>
        <div className={classes.StateItem}>
          {p.kind === 'ADMISSION'
            ? props.t('examSession.registration')
            : props.t('examSession.registration.postAdmission')}
        </div>
        <div className={classes.FirstShowOnHover}>
          {p.state === 'COMPLETED' && !props.disableControls
            ? relocateParticipant(p)
            : null}
        </div>
        <div className={classes.Item} />
        <div className={classes.Item}>{ssnOrBirthDate(p.form)}</div>
        <div className={classes.Item}>
          {p.form.street_address} {p.form.zip}
          {', '}
          {p.form.post_office}
        </div>
        <div className={classes.Item}>{getPhoneNumber(p)}</div>
        <div className={classes.Item}> {p.form.email}</div>
        <div className={classes.ShowOnHover}>
          {(p.state === 'SUBMITTED' || p.state === 'COMPLETED') &&
          !props.disableControls
            ? cancelRegistrationButton(p)
            : null}
        </div>
        <span className={classes.Line} />
        <span className={classes.LineEnd} />
      </React.Fragment>
    ));
  };

  const participantsHeader = () => {
    const participantsCount = examSessionParticipantsCount(props.examSession);
    return (
      <h2>
        {props.t('examSession.participants')}
        {':'} {participantsCount.participants} /{' '}
        {participantsCount.maxParticipants}
      </h2>
    );
  };
  return (
    <div data-cy="participant-list">
      {participantsHeader()}

      {props.examSession.queue > 0 && (
        <h3>
          {props.t('examSession.inQueue')}
          {':'} {props.examSession.queue}
        </h3>
      )}

      {props.participants.length > 0 && (
        <React.Fragment>
          <div className={classes.ListExport}>
            <ListExport participants={props.participants} />
            {participantFiltering()}
          </div>
          <div className={classes.ParticipantList}>
            {participantRows(props.participants)}
          </div>
        </React.Fragment>
      )}
    </div>
  );
};

participantList.propTypes = {
  examSession: PropTypes.object.isRequired,
  examSessions: PropTypes.array.isRequired,
  participants: PropTypes.array.isRequired,
  onCancel: PropTypes.func.isRequired,
  onConfirmPayment: PropTypes.func.isRequired,
  onRelocate: PropTypes.func.isRequired,
  onResendLink: PropTypes.func.isRequired,
  disableControls: PropTypes.bool,
};

export default withTranslation()(participantList);
