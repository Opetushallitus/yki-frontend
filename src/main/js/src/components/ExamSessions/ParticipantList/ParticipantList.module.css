.ParticipantList {
  display: grid;
  grid-template-columns: 20px 1fr 2fr 1fr 1fr 1fr;
  grid-gap: 4px;
  grid-row-gap: 0px;
}

.ListExport {
  margin-bottom: 8px;
}

.Item {
  font-size: 0.875em;
}

.StateItem {
  font-size: 0.875em;
}

.ItemHeader {
  font-size: 0.875em;
  font-weight: bold;
  white-space: nowrap;
}

.Status {
  color: hsl(15, 100%, 40%);
}

.StatusCompleted {
  color: hsl(96, 96%, 27%);
}

.Index {
  text-align: right;
  font-weight: bold;
}

.Line {
  grid-column: 1/6;
  margin-top: 4px;
  margin-bottom: 8px;
  border-bottom: 1px solid hsl(0, 0%, 80%);
}

.LineEnd {
  grid-column: 6/7;
  margin-top: 4px;
  margin-bottom: 8px;
}

.ResendEmailLink {
  font-size: 8px;
}

.ResendEmailButton {
  margin-left: 5px;
}

.ResendEmailSelectionText {
  font-size: 8;
  font-weight: normal;
  color: rgb(41, 41, 41);
}

.ParticipantFilter {
  margin-left: 5px;
  width: 200px;
  height: 28px;
  border: 1px solid hsl(0, 0%, 68%);
  border-radius: 2px;
  background-color: hsl(0, 0%, 100%);
}

.ShowOnHover {
  opacity: 0;
}

.ShowOnHover:hover {
  opacity: 1;
}

.Item:hover ~ .ShowOnHover {
  opacity: 1;
}

.Item:hover ~ .ShowOnHover + .Line ~ .ShowOnHover {
  opacity: 0;
}

.FirstShowOnHover {
  opacity: 0;
}

.FirstShowOnHover:hover {
  opacity: 1;
}

.StateItem:hover ~ .FirstShowOnHover {
  opacity: 1;
}

.StateItem:hover ~ .ShowOnHover + .Line ~ .FirstShowOnHover {
  opacity: 0;
}
