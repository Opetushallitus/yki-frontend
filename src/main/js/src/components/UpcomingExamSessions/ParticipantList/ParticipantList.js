import React from 'react';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';
import moment from 'moment';
import { parsePhoneNumberFromString } from 'libphonenumber-js';

import { DATE_FORMAT } from '../../../common/Constants';
import checkMarkDone from '../../../assets/svg/checkmark-done.svg';
import checkMarkNotDone from '../../../assets/svg/checkmark-not-done.svg';
import trashcan from '../../../assets/svg/trashcan.svg';
import classes from './ParticipantList.module.css';
import { ActionButton } from '../../UI/ActionButton/ActionButton';

export const participantList = props => {
  const registratioStatus = registrationState => {
    const image =
      registrationState === 'COMPLETED' ? checkMarkDone : checkMarkNotDone;
    const text =
      registrationState === 'COMPLETED'
        ? props.t('examSession.paid')
        : props.t('examSession.notPaid');
    return (
      <React.Fragment>
        <img src={image} alt="" /> {text}
      </React.Fragment>
    );
  };

  const ssnOrBirthDate = form => {
    return form.ssn ? form.ssn : moment(form.birthdate).format(DATE_FORMAT);
  };

  const cancelRegistration = () => {
    return (
      <React.Fragment>
        <img src={trashcan} alt="" />{' '}
        {props.t('examSession.registration.cancel')}
      </React.Fragment>
    );
  };

  const participantRows = participants => {
    return participants.map((p, i) => (
      <React.Fragment key={i}>
        <div
          className={[classes.ItemHeader, classes.Index].join(' ')}
          data-cy={`participant-${p.registration_id}`}
        >
          {i + 1}.
        </div>
        <div className={classes.ItemHeader}>
          {p.form.first_name} {p.form.last_name}
        </div>
        <div className={[classes.ItemHeader, classes.Status].join(' ')}>
          {registratioStatus(p.state)}
        </div>
        <div className={classes.Item}/>
        <div className={classes.Item}/>
        <div className={classes.Item}/>
        <div className={classes.Item}/>
        <div className={classes.Item}>{ssnOrBirthDate(p.form)}</div>
        <div className={classes.Item}>
          {p.form.street_address} {p.form.zip}
          {', '}
          {p.form.post_office}
        </div>
        <div className={classes.Item}>
          {parsePhoneNumberFromString(
            p.form.phone_number,
          ).formatInternational()}
        </div>
        <div className={classes.Item}> {p.form.email}</div>
        <div className={[classes.ShowOnHover].join(' ')}>
          <ActionButton
            children={cancelRegistration()}
            onClick={() =>
              props.onCancel(
                props.examSession.organizer_oid,
                props.examSession.id,
                p.registration_id,
              )
            }
            confirmText={props.t('examSession.registration.cancel.confirm')}
            cancelText={props.t('examSession.registration.cancel.cancel')}
          />
        </div>
        <span className={classes.Line} />
        <span className={classes.LineEnd} />
      </React.Fragment>
    ));
  };

  return (
    <div data-cy="participant-list">
      <h3>
        {props.t('examSession.participants')}
        {':'} {props.participants.length} / {props.examSession.max_participants}
      </h3>
      <div className={classes.ParticipantList}>
        {participantRows(props.participants)}
      </div>
    </div>
  );
};

participantList.propTypes = {
  examSession: PropTypes.object.isRequired,
  participants: PropTypes.array.isRequired,
  onCancel: PropTypes.func.isRequired,
};

export default withTranslation()(participantList);
