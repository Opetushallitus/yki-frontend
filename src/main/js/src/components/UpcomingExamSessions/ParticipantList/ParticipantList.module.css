.ParticipantList {
  display: grid;
  grid-template-columns: 20px 1fr 2fr 1fr 1fr 1fr;
  grid-gap: 4px;
  grid-row-gap: 0px;
}

.Item {
  font-size: 0.875em;
}

.FirstLineItem {
  font-size: 0.875em;
}


.ItemHeader {
  display: grid;
  font-size: 0.875em;
  font-weight: bold;
  white-space: nowrap;
}

.Status {
  display: inline-block;
  color: hsla(96, 96%, 27%, 1);
}

.Index {
  text-align: right;
  font-weight: bold;
}

.Line {
  grid-column: 1/6;
  margin-top: 3px;
  margin-bottom: 8px;
  border-bottom: 1px solid hsla(0, 0%, 80%, 1);
}

.LineEnd {
  grid-column: 6/7;
  margin-top: 3px;
  margin-bottom: 8px;
}

.ShowOnHover {
  opacity: 0.0;
}

.ShowOnHover:hover {
  opacity: 1.0;
}

.Item:hover ~ .ShowOnHover {
  opacity: 1.0;
}

.Item:hover ~ .ShowOnHover + .Line ~ .ShowOnHover {
  opacity: 0.0;
}

.FirstShowOnHover {
  opacity: 0.0;
}

.FirstShowOnHover:hover {
  opacity: 1.0;
}

.FirstLineItem:hover ~ .FirstShowOnHover {
  opacity: 1.0;
}

.FirstLineItem:hover ~ .ShowOnHover + .Line ~ .FirstShowOnHover {
  opacity: 0.0;
}
