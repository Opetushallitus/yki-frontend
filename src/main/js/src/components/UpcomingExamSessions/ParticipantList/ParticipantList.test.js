import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import { participantList as ParticipantList } from './ParticipantList';

configure({ adapter: new Adapter() });

const examSessionFirst = {
  registration_end_date: '2028-12-15',
  session_date: '2028-12-30',
  participants: 2,
  max_participants: 50,
  registration_start_date: '2018-09-01',
  language_code: 'eng',
  level_code: 'KESKI',
};

const examSessionSecond = {
  ...examSessionFirst,
  session_date: '2029-01-01',
};

const participants = [
  {
    form: {
      post_office: 'Espoo',
      ssn: '011088-1234 ',
      phone_number: '+358405131441',
      email: 'michael.wallis@test.com',
      last_name: 'Michael',
      first_name: 'Wallis',
      street_address: 'Revontulentie 13 C 2',
      zip: '02100',
    },
    state: 'COMPLETED',
    registration_id: 1,
  },
  {
    form: {
      post_office: 'Espoo',
      ssn: '011088-1234 ',
      phone_number: '+358405131441',
      email: 'natalia.gaddens@test.com',
      last_name: 'Natalia',
      first_name: 'Gaddens',
      street_address: 'Revontulentie 13 C 2',
      zip: '02100',
    },
    state: 'SUBMITTED',
    registration_id: 2,
  },
];

describe('<ParticipantList />', () => {
  it('should render participant rows', () => {
    const wrapper = shallow(
      <ParticipantList
        examSession={examSessionFirst}
        examSessions={[examSessionFirst, examSessionSecond]}
        participants={participants}
        t={t => t}
        onCancel={jest.fn()}
        onConfirmPayment={jest.fn()}
        onRelocate={jest.fn()}
        onResendLink={jest.fn()}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
