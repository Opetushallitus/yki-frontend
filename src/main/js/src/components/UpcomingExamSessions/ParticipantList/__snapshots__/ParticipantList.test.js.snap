// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ParticipantList /> should render participant rows 1`] = `
<div
  data-cy="participant-list"
>
  <h3>
    examSession.participants
    :
     
    2
     / 
    50
  </h3>
  <div
    className="ParticipantList"
  >
    <div
      className="ItemHeader Index StateItem"
      data-cy="participant-1"
    >
      1
      .
    </div>
    <div
      className="ItemHeader StateItem"
    >
      Wallis
       
      Michael
    </div>
    <div
      className="ItemHeader Status StateItem"
    >
      <img
        alt=""
        data-cy="registration-COMPLETED"
        src="checkmark-done.svg"
      />
       
      examSession.paid
    </div>
    <div
      className="StateItem"
    />
    <div
      className="StateItem"
    />
    <div
      className="FirstShowOnHover"
    >
      <ActionButton
        cancelText="examSession.registration.confirmPayment.cancel"
        confirmText="examSession.registration.confirmPayment.confirm"
        onClick={[Function]}
      />
    </div>
    <div
      className="Item"
    />
    <div
      className="Item"
    >
      011088-1234 
    </div>
    <div
      className="Item"
    >
      Revontulentie 13 C 2
       
      02100
      , 
      Espoo
    </div>
    <div
      className="Item"
    >
      +358 40 5131441
    </div>
    <div
      className="Item"
    >
       
      michael.wallis@test.com
    </div>
    <div
      className="ShowOnHover"
    >
      <ActionButton
        cancelText="examSession.registration.cancel.cancel"
        confirmText="examSession.registration.cancel.confirm"
        onClick={[Function]}
      >
        <img
          alt=""
          src="trashcan.svg"
        />
         
        examSession.registration.cancel
      </ActionButton>
    </div>
    <span
      className="Line"
    />
    <span
      className="LineEnd"
    />
    <div
      className="ItemHeader Index StateItem"
      data-cy="participant-2"
    >
      2
      .
    </div>
    <div
      className="ItemHeader StateItem"
    >
      Gaddens
       
      Natalia
    </div>
    <div
      className="ItemHeader Status StateItem"
    >
      <img
        alt=""
        data-cy="registration-SUBMITTED"
        src="checkmark-not-done.svg"
      />
       
      examSession.notPaid
    </div>
    <div
      className="StateItem"
    />
    <div
      className="StateItem"
    />
    <div
      className="FirstShowOnHover"
    >
      <ActionButton
        cancelText="examSession.registration.confirmPayment.cancel"
        confirmText="examSession.registration.confirmPayment.confirm"
        onClick={[Function]}
      >
        <img
          alt=""
          data-cy="confirm-payment-icon"
          src="checkmark-done.svg"
        />
         
        examSession.registration.confirmPayment
      </ActionButton>
    </div>
    <div
      className="Item"
    />
    <div
      className="Item"
    >
      011088-1234 
    </div>
    <div
      className="Item"
    >
      Revontulentie 13 C 2
       
      02100
      , 
      Espoo
    </div>
    <div
      className="Item"
    >
      +358 40 5131441
    </div>
    <div
      className="Item"
    >
       
      natalia.gaddens@test.com
    </div>
    <div
      className="ShowOnHover"
    >
      <ActionButton
        cancelText="examSession.registration.cancel.cancel"
        confirmText="examSession.registration.cancel.confirm"
        onClick={[Function]}
      >
        <img
          alt=""
          src="trashcan.svg"
        />
         
        examSession.registration.cancel
      </ActionButton>
    </div>
    <span
      className="Line"
    />
    <span
      className="LineEnd"
    />
  </div>
</div>
`;
